Tight coupling of kafka version

## Describe the bug

There is a tight coupling of the kafka version between the this project and components that use it.  This project has a dependency on several kafka components, which becomes a transitive dependency in the user's projects. 

We saw the situation below where the user's project had a dependency on a single kafka artefact (kafka-client) at a different version.  This lead to a classpath stramash and a runtime failure as an expected symbol was missing.

```
[ERROR] io.kroxylicious.test.tester.KroxyliciousTestersTest.shouldReturnClosableProducer(KafkaCluster) -- Time elapsed: 0.002 s <<< ERROR!
> org.junit.jupiter.api.extension.ParameterResolutionException: Failed to resolve parameter 
```

## Workaround

The user could use dependency management overrides to control all the kafka artefacts to the same version to workaround.  It is a bit opaque as the user wouldn't be entirely certain which artefacts to dependency managed. 

## Proposed Solution

It would probably be better if this project used `provided` dependencies on the kafka artefacts, and insist the user define the kafka version they want to use.  The documentation for this component would need to call out which artefacts are required (kafka-server, -storage etc).
 
              > Upgrading to a version of the junit extension (0.6.0-SNAPSHOT) fixes the parameter resolution error. Lets get the extension released and then revisit the tight coupling of versions.
> 
> ```
> [ERROR] io.kroxylicious.test.tester.KroxyliciousTestersTest.shouldReturnClosableProducer(KafkaCluster) -- Time elapsed: 0.002 s <<< ERROR!
> org.junit.jupiter.api.extension.ParameterResolutionException: Failed to resolve parameter [io.kroxylicious.testing.kafka.api.KafkaCluster cluster] in method [void io.kroxylicious.test.tester.KroxyliciousTestersTest.shouldReturnClosableProducer(io.kroxylicious.testing.kafka.api.KafkaCluster)]: Could not initialize class kafka.server.KafkaConfig$
> 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
> 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
> Caused by: java.lang.NoClassDefFoundError: Could not initialize class kafka.server.KafkaConfig$
> 	at kafka.server.KafkaConfig.LogDirProp(KafkaConfig.scala)
> 	at io.kroxylicious.testing.kafka.invm.InVMKafkaCluster.buildBrokerConfig(InVMKafkaCluster.java:165)
> 	at io.kroxylicious.testing.kafka.invm.InVMKafkaCluster.buildKafkaServer(InVMKafkaCluster.java:106)
> 	at io.kroxylicious.testing.kafka.invm.InVMKafkaCluster.lambda$start$5(InVMKafkaCluster.java:186)
> 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
> 	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
> 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
> 	at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
> 	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)
> 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
> 	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:667)
> 	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)
> 	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)
> 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
> 	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
> 	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:765)
> 	at io.kroxylicious.testing.kafka.invm.InVMKafkaCluster.start(InVMKafkaCluster.java:185)
> 	at io.kroxylicious.testing.kafka.junit5ext.KafkaClusterExtension.createCluster(KafkaClusterExtension.java:956)
> 	at io.kroxylicious.testing.kafka.junit5ext.KafkaClusterExtension.lambda$getCluster$12(KafkaClusterExtension.java:814)
> 	at io.kroxylicious.testing.kafka.junit5ext.KafkaClusterExtension.getCluster(KafkaClusterExtension.java:812)
> 	at io.kroxylicious.testing.kafka.junit5ext.KafkaClusterExtension.resolveParameter(KafkaClusterExtension.java:492)
> 	at io.kroxylicious.testing.kafka.junit5ext.KafkaClusterExtension.resolveParameter(KafkaClusterExtension.java:464)
> 	... 2 more
> Caused by: java.lang.ExceptionInInitializerError: Exception java.lang.NoSuchFieldError: id [in thread "main"]
> 	at kafka.server.Defaults$.<clinit>(KafkaConfig.scala:186)
> 	at kafka.server.KafkaConfig$.<clinit>(KafkaConfig.scala:734)
> 	... 24 more
> ```


_Originally posted by @k-wall in https://github.com/kroxylicious/kroxylicious/issues/665#issuecomment-1754700649_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,15 @@ Please enumerate all user-facing changes using format `<githib issue/pr number>:
 ## 1.0.0
 
 
+## 0.7.0
+
+* [#197](https://github.com/kroxylicious/kroxylicious-junit5-extension/issues/197): Consumer of the extension should be able to choose the version of the kafka dependency.
+
+### Changes, deprecations and removals
+
+* It is now necessary to declare org.apache.kafka:kafka_2.13 as a test dependency.  If you wish to use kafka in-vm with zookeeper, org.apache.zookeeper:zookeeper
+  must be a test dependency too.
+
 
 ## 0.6.0
 

--- a/README.md
+++ b/README.md
@@ -16,11 +16,27 @@ testImplementation 'io.kroxylicious.testing:testing-junit5-extension:0.1'
 Maven
 
 ```xml
-<dependency>
-    <groupId>io.kroxylicious.testing</groupId>
-    <artifactId>testing-junit5-extension</artifactId>
-    <version>0.1</version>
-</dependency>
+<dependencies>
+    <dependency>
+        <groupId>io.kroxylicious.testing</groupId>
+        <artifactId>testing-junit5-extension</artifactId>
+        <version>0.1</version>
+    </dependency>
+    <!-- Kafka Broker version you want to use -->
+    <dependency>
+        <groupId>org.apache.kafka</groupId>
+        <artifactId>kafka_2.13</artifactId>
+        <version>${kafka.version}</version> <!-- versions 3.3.0 or higher is known to work -->
+        <scope>test</scope>
+    </dependency>
+    <!-- Optional, required if you want to use in-VM kafka in Zookeeper mode. -->
+    <dependency>
+        <groupId>org.apache.zookeeper</groupId>
+        <artifactId>zookeeper</artifactId>
+        <version>${zookeeper.version}</version> <!-- version 3.6.3 or higher is known to work -->
+        <scope>test</scope>
+    </dependency>
+</dependencies>
 ```
 
 ## Example

--- a/impl/pom.xml
+++ b/impl/pom.xml
@@ -42,10 +42,6 @@
             <artifactId>kafka_2.13</artifactId>
         </dependency>
         <dependency>
-            <groupId>org.apache.kafka</groupId>
-            <artifactId>kafka-server-common</artifactId>
-        </dependency>
-        <dependency>
             <groupId>org.apache.zookeeper</groupId>
             <artifactId>zookeeper</artifactId>
         </dependency>

--- a/integration-test/pom.xml
+++ b/integration-test/pom.xml
@@ -43,10 +43,6 @@
             <artifactId>kafka_2.13</artifactId>
         </dependency>
         <dependency>
-            <groupId>org.apache.kafka</groupId>
-            <artifactId>kafka-server-common</artifactId>
-        </dependency>
-        <dependency>
             <groupId>org.apache.zookeeper</groupId>
             <artifactId>zookeeper</artifactId>
         </dependency>

--- a/junit5-extension/pom.xml
+++ b/junit5-extension/pom.xml
@@ -39,7 +39,7 @@
         </dependency>
         <dependency>
             <groupId>org.apache.kafka</groupId>
-            <artifactId>kafka-clients</artifactId>
+            <artifactId>kafka_2.13</artifactId>
         </dependency>
         <dependency>
             <groupId>org.junit.jupiter</groupId>

--- a/pom.xml
+++ b/pom.xml
@@ -175,7 +175,8 @@
                 <!-- The user needs to provide this dependency only if they wish to use in-VM kafka with a Zookeeper
                      controller.  If they are using in-VM kafka in KRaft node, or container kafka with Zookeeper, it is not
                      required. -->
-                <optional>true</optional>
+                <scope>provided</scope>
+                <optional>true</optional> <!-- MNG-5227 -->
             </dependency>
             <dependency>
                 <groupId>org.awaitility</groupId>

--- a/pom.xml
+++ b/pom.xml
@@ -155,12 +155,6 @@
                 <scope>provided</scope>
             </dependency>
             <dependency>
-                <groupId>org.apache.kafka</groupId>
-                <artifactId>kafka-server-common</artifactId>
-                <version>${kafka.version}</version>
-                <scope>provided</scope>
-            </dependency>
-            <dependency>
                 <groupId>org.apache.zookeeper</groupId>
                 <artifactId>zookeeper</artifactId>
                 <version>${zookeeper.version}</version>

--- a/pom.xml
+++ b/pom.xml
@@ -172,7 +172,10 @@
                         <artifactId>logback-classic</artifactId>
                     </exclusion>
                 </exclusions>
-                <scope>provided</scope>
+                <!-- The user needs to provide this dependency only if they wish to use in-VM kafka with a Zookeeper
+                     controller.  If they are using in-VM kafka in KRaft node, or container kafka with Zookeeper, it is not
+                     required. -->
+                <optional>true</optional>
             </dependency>
             <dependency>
                 <groupId>org.awaitility</groupId>

--- a/pom.xml
+++ b/pom.xml
@@ -131,6 +131,7 @@
                 <groupId>org.apache.kafka</groupId>
                 <artifactId>kafka-clients</artifactId>
                 <version>${kafka.version}</version>
+                <scope>provided</scope>
             </dependency>
             <dependency>
                 <groupId>org.apache.kafka</groupId>
@@ -151,11 +152,13 @@
                         <artifactId>logback-classic</artifactId>
                     </exclusion>
                 </exclusions>
+                <scope>provided</scope>
             </dependency>
             <dependency>
                 <groupId>org.apache.kafka</groupId>
                 <artifactId>kafka-server-common</artifactId>
                 <version>${kafka.version}</version>
+                <scope>provided</scope>
             </dependency>
             <dependency>
                 <groupId>org.apache.zookeeper</groupId>
@@ -175,6 +178,7 @@
                         <artifactId>logback-classic</artifactId>
                     </exclusion>
                 </exclusions>
+                <scope>provided</scope>
             </dependency>
             <dependency>
                 <groupId>org.awaitility</groupId>

